src/lem/circuit.rs

```diff
@@ -0,0 +1,904 @@
+    #[inline]
+    pub(crate) fn consume_hash4(&mut self) -> usize {
+        self.hash4 += 1;
+        self.hash4 - 1
+    }
+
+    #[inline]
+    pub(crate) fn max(&self, other: Self) -> Self {
+        use std::cmp::max;
+        Self {
+            hash2: max(self.hash2, other.hash2),
+            hash3: max(self.hash3, other.hash3),
+            hash4: max(self.hash4, other.hash4),
+        }
+    }
+
+    #[inline]
+    pub(crate) fn add(&self, other: Self) -> Self {
+        Self {
+            hash2: self.hash2 + other.hash2,
+            hash3: self.hash3 + other.hash3,
+            hash4: self.hash4 + other.hash4,
+        }
+    }
+}
+
+impl LEMCTL {
+    pub fn count_slots(&self) -> SlotsCounter {
+        match self {
+            LEMCTL::MatchTag(_, cases) => {
+                cases.values().fold(SlotsCounter::default(), |acc, block| {
+                    acc.max(block.count_slots())
+                })
+            }
+            LEMCTL::MatchSymbol(_, cases, def) => cases
+                .values()
+                .fold(def.count_slots(), |acc, block| acc.max(block.count_slots())),
+            LEMCTL::Return(..) => SlotsCounter::default(),
+            LEMCTL::Seq(ops, rest) => {
+                let ops_slots = ops.iter().fold(SlotsCounter::default(), |acc, op| {
+                    let val = match op {
+                        LEMOP::Hash2(..) | LEMOP::Unhash2(..) => SlotsCounter::new((1, 0, 0)),
+                        LEMOP::Hash3(..) | LEMOP::Unhash3(..) => SlotsCounter::new((0, 1, 0)),
+                        LEMOP::Hash4(..) | LEMOP::Unhash4(..) => SlotsCounter::new((0, 0, 1)),
+                        _ => SlotsCounter::default(),
+                    };
+                    acc.add(val)
+                });
+                ops_slots.add(rest.count_slots())
+            }
+        }
+    }
+}
+
+/// Manages global allocations for constants in a constraint system
+#[derive(Default)]
+pub(crate) struct GlobalAllocator<F: LurkField>(HashMap<FWrap<F>, AllocatedNum<F>>);
+
+#[inline]
+fn allocate_num<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    namespace: &str,
+    value: F,
+) -> Result<AllocatedNum<F>> {
+    AllocatedNum::alloc(cs.namespace(|| namespace), || Ok(value))
+        .with_context(|| format!("allocation for '{namespace}' failed"))
+}
+
+impl<F: LurkField> GlobalAllocator<F> {
+    /// Checks if the allocation for a numeric variable has already been cached.
+    /// If so, return the cached allocation variable. Allocate, cache and return
+    /// otherwise.
+    pub(crate) fn get_or_alloc_num<CS: ConstraintSystem<F>>(
+        &mut self,
+        cs: &mut CS,
+        f: F,
+    ) -> Result<AllocatedNum<F>> {
+        let wrap = FWrap(f);
+        match self.0.get(&wrap) {
+            Some(allocated_num) => Ok(allocated_num.to_owned()),
+            None => {
+                let allocated_num =
+                    allocate_num(cs, &format!("globally allocate {}", f.hex_digits()), f)?;
```

Comment:
This appears to be an unconstrained allocation (calling `AllocatedNum::alloc()` via `allocate_num()` above.

I think what you actually want may be `circuit::gadgets::data::allocate_constant`.

Reply:
Changed